# 1.读取数据

In [1]:
import pandas as pd

D:\ADAnZh\lib\site-packages\pandas\core\computation\expressions.py:20: UserWarning: Pandas requires version '2.7.3' or newer of 'numexpr' (version '2.7.1' currently installed).
  from pandas.core.computation.check import NUMEXPR_INSTALLED


In [2]:
original_data = pd.read_csv("E_commerce.csv")

<ipython-input-2-5450507f7e84>:1: DtypeWarning: Columns (18,19) have mixed types. Specify dtype option on import or set low_memory=False.
  original_data = pd.read_csv("E_commerce.csv")


### 大概看一下数据长什么样

In [3]:
original_data.head()

Customer ID    Customer Name      Segment        City       State  \
0      LS-001     Lane Daniels     Consumer    Brisbane  Queensland   
1      IZ-002    Alvarado Kriz  Home Office      Berlin      Berlin   
2      EN-003       Moon Weien     Consumer     Porirua  Wellington   
3      AN-004  Sanchez Bergman    Corporate       Kabul       Kabul   
4      ON-005     Rowe Jackson    Corporate  Townsville  Queensland   

       Country        Region Gender  Age         Education  ...     Sales  \
0    Australia       Oceania   Male   22  Associate Degree  ...  $140.00    
1      Germany       Central   Male   32          Bachelor  ...  $211.00    
2  New Zealand       Oceania   Male   21       High School  ...  $117.00    
3  Afghanistan  Central Asia   Male   22          Bachelor  ...  $118.00    
4    Australia       Oceania   Male   28  Associate Degree  ...  $250.00    

  Quantity Discount    Profit Shipping Cost Order Priority  \
0        2     0.05   $46.00         $4.60          Medium   
1        3     0.03  $112.00        $11.20          Medium   
2        5     0.01   $31.20         $3.10        Critical   
3        2     0.05   $26.20         $2.60            High   
4        1     0.04  $160.00        $16.00        Critical   

  Browsing Time (min) Like Share Add to Cart  
0                14.7    1     1           1  
1                15.0    0     0           1  
2                19.9    1     1           1  
3                15.8    1     1           1  
4                18.1    1     1           1  

[5 rows x 27 columns]

In [4]:
# 临时设置并显示
with pd.option_context("display.max_columns", 150):
    print(original_data)
original_data

      Customer ID    Customer Name      Segment        City          State  \
0          LS-001     Lane Daniels     Consumer    Brisbane     Queensland   
1          IZ-002    Alvarado Kriz  Home Office      Berlin         Berlin   
2          EN-003       Moon Weien     Consumer     Porirua     Wellington   
3          AN-004  Sanchez Bergman    Corporate       Kabul          Kabul   
4          ON-005     Rowe Jackson    Corporate  Townsville     Queensland   
...           ...              ...          ...         ...            ...   
51284    IN-00783       Welch Fein    Corporate     Baghdad        Baghdad   
51285    TT-00135  Martinez Arnett    Corporate      Kanpur  Uttar Pradesh   
51286    ON-00828     Mccoy Duston  Home Office    Portland         Oregon   
51287     RN-0016   Bentley Zypern     Consumer      Berlin         Berlin   
51288   RZ-001718  Mcclure Schwarz  Home Office   La Romana      La Romana   

                  Country        Region  Gender  Age         Ed

Customer ID    Customer Name      Segment        City          State  \
0          LS-001     Lane Daniels     Consumer    Brisbane     Queensland   
1          IZ-002    Alvarado Kriz  Home Office      Berlin         Berlin   
2          EN-003       Moon Weien     Consumer     Porirua     Wellington   
3          AN-004  Sanchez Bergman    Corporate       Kabul          Kabul   
4          ON-005     Rowe Jackson    Corporate  Townsville     Queensland   
...           ...              ...          ...         ...            ...   
51284    IN-00783       Welch Fein    Corporate     Baghdad        Baghdad   
51285    TT-00135  Martinez Arnett    Corporate      Kanpur  Uttar Pradesh   
51286    ON-00828     Mccoy Duston  Home Office    Portland         Oregon   
51287     RN-0016   Bentley Zypern     Consumer      Berlin         Berlin   
51288   RZ-001718  Mcclure Schwarz  Home Office   La Romana      La Romana   

                  Country        Region  Gender  Age         Education  ...  \
0               Australia       Oceania    Male   22  Associate Degree  ...   
1                 Germany       Central    Male   32          Bachelor  ...   
2             New Zealand       Oceania    Male   21       High School  ...   
3             Afghanistan  Central Asia    Male   22          Bachelor  ...   
4               Australia       Oceania    Male   28  Associate Degree  ...   
...                   ...           ...     ...  ...               ...  ...   
51284                Iraq          EMEA  Female   20  Associate Degree  ...   
51285               India  Central Asia    Male   20       High School  ...   
51286       United States          West    Male   33            Master  ...   
51287             Germany       Central  Female   48         Doctorate  ...   
51288  Dominican Republic     Caribbean  Female   36            Master  ...   

          Sales Quantity Discount    Profit Shipping Cost Order Priority  \
0      $140.00         2     0.05   $46.00         $4.60          Medium   
1      $211.00         3     0.03  $112.00        $11.20          Medium   
2      $117.00         5     0.01   $31.20         $3.10        Critical   
3      $118.00         2     0.05   $26.20         $2.60            High   
4      $250.00         1     0.04  $160.00        $16.00        Critical   
...         ...      ...      ...       ...           ...            ...   
51284   $85.00         5     0.04   $17.00         $1.70          Medium   
51285   $85.00         1     0.03    $2.50         $0.20          Medium   
51286   $85.00         1     0.05    $0.80         $0.10          Medium   
51287   $85.00         3     0.04   $28.30         $2.80          Medium   
51288   $85.00         3     0.03   $28.30         $2.80          Medium   

      Browsing Time (min) Like Share Add to Cart  
0                    14.7    1     1           1  
1                    15.0    0     0           1  
2                    19.9    1     1           1  
3                    15.8    1     1           1  
4                    18.1    1     1           1  
...                   ...  ...   ...         ...  
51284                21.4    1     1           1  
51285                15.5    0     0           1  
51286                14.3    0     0           1  
51287                 8.4    1     1           1  
51288                25.8    1     1           1  

[51289 rows x 27 columns]

In [5]:
#pd才有set_option的方法
#pd.set_option()是直接设置 Pandas 的显示配置，不会返回数据框
#正确做法
#import pandas as pd

# # 首先设置显示选项（这会影响所有DataFrame的显示）
# pd.set_option("display.max_columns", 150)

# # 然后读取或创建您的数据框
# original_data = pd.read_csv("your_file.csv")  # 或其他数据加载方式

# 如果先读了csv文件，临时展示# 临时设置并显示
# with pd.option_context("display.max_columns", 150):
#     print(original_data)
# original_data

# original_data = pd.set_option("display.max_columns",150)#pd才有set_option的方法
# original_data


# 2.评估数据

在这一部分,我将对在上一部分建立的original_data这个DataFrame所包含的数据进行评估。

评估主要从两个方面进行:
### 结构和内容,
即整齐度和干净度。数据的结构性问题指不符合
### “每列是一个变量,每行是一个观察值,每个单元格是一个值”这三个标准,数据的内容性问题包括存在丢失数据、重复数据、无效数据等。

评估数据整齐度

In [6]:
original_data.sample(10)

Customer ID     Customer Name      Segment            City  \
8465     LD-00222       Obrien Geld    Corporate           Prato   
24073    RT-00464        Bowen Hart     Consumer           Paris   
49909    NK-00853     Randolph Sink  Home Office           Perth   
19474   ER-001215       Best Venier     Consumer            Troy   
37919     TT-0049     Fuller Eplett    Corporate       San Diego   
49182    OV-00481      Waller Yotov    Corporate   South Shields   
40835    NG-00596   Mccall Smayling     Consumer     Los Angeles   
24251   RD-004435    Dalton Radford     Consumer      Gold Coast   
23844    RR-00218  Underwood Mcgarr  Home Office  Stoke-on-Trent   
41257    ER-00679       Hardy Greer    Corporate        Shanghai   

                   State         Country      Region  Gender  Age  \
8465             Tuscany           Italy       South  Female   18   
24073      Ile-de-France          France     Central    Male   26   
49909  Western Australia       Australia     Oceania    Male   20   
19474           New York   United States        East  Female   30   
37919         California   United States        West  Female   28   
49182            England  United Kingdom       North    Male   44   
40835         California   United States        West  Female   22   
24251         Queensland       Australia     Oceania  Female   35   
23844            England  United Kingdom       North    Male   21   
41257           Shanghai           China  North Asia    Male   25   

              Education  ...     Sales Quantity Discount    Profit  \
8465        High School  ...  $104.00         4     0.04    $7.40    
24073          Bachelor  ...  $224.00         2     0.04  $126.10    
49909       High School  ...  $122.00         3     0.01   $38.30    
19474          Bachelor  ...  $216.00         2     0.01  $131.70    
37919  Associate Degree  ...  $109.00         5     0.04    $7.20    
49182            Master  ...   $85.00         4     0.01    $1.60    
40835       High School  ...   $85.00         3     0.02   $28.30    
24251         Doctorate  ...   $62.00         3     0.01   $20.70    
23844       High School  ...   $62.00         1     0.02   $62.00    
41257       High School  ...   $62.00         4     0.04   $15.50    

      Shipping Cost Order Priority Browsing Time (min) Like Share Add to Cart  
8465         $0.70            High                 4.7    0     1           0  
24073       $12.60          Medium                11.5    1     1           0  
49909        $3.80          Medium                10.7    0     0           0  
19474       $13.20          Medium                30.9    1     1           1  
37919        $0.70          Medium                19.7    1     1           1  
49182        $0.20          Medium                16.8    1     0           0  
40835        $2.80          Medium                23.1    1     0           1  
24251        $2.10            High                15.1    1     1           1  
23844        $6.20             Low                17.3    0     0           1  
41257        $1.60            High                15.5    0     0           0  

[10 rows x 27 columns]

从抽样的10行数据数据来看,数据符合“每列是一个变量,每行是一个观察值,每个单元格是一个值”,具体来看每行是关于某商
品的一次交易,每列是交易相关的各个变量,因此不存在结构性问题。

### 评估数据干净度

In [7]:
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51289 entries, 0 to 51288
Data columns (total 27 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Customer ID          51289 non-null  object 
 1   Customer Name        51289 non-null  object 
 2   Segment              51289 non-null  object 
 3   City                 51289 non-null  object 
 4   State                51289 non-null  object 
 5   Country              51289 non-null  object 
 6   Region               51215 non-null  object 
 7   Gender               51289 non-null  object 
 8   Age                  51289 non-null  int64  
 9   Education            51289 non-null  object 
 10  Marital Status       51289 non-null  object 
 11  Order ID             51289 non-null  object 
 12  Order Date           51289 non-null  object 
 13  Months               51289 non-null  object 
 14  Ship Mode            51289 non-null  object 
 15  Product Category     51289 non-null 

从输出结果来看,数据共有51289条观察值,而Order Priority变量51287存在缺失值。

此外,Browsing Time (min)的数据类型应为日期应当进行数据格式转换。

### 评估缺失数据

在了解 Order Priority 存在缺失值后,根据条件提取出缺失观察值。


In [8]:
original_data['Order Priority'].isnull()

0        False
1        False
2        False
3        False
4        False
         ...  
51284    False
51285    False
51286    False
51287    False
51288    False
Name: Order Priority, Length: 51289, dtype: bool

In [9]:
original_data[original_data['Order Priority'].isnull()]

Customer ID  Customer Name   Segment               City             State  \
624    TT-00626   Harper Dartt  Consumer           Coyoacán  Distrito Federal   
790    IS-00792  Navarro Preis  Consumer  Ciudad del Carmen          Campeche   

    Country Region Gender  Age    Education  ...     Sales Quantity Discount  \
624  Mexico  North   Male   18  High School  ...  $250.00         4     0.03   
790  Mexico  North   Male   36       Master  ...  $231.00         1     0.01   

       Profit Shipping Cost Order Priority Browsing Time (min) Like Share  \
624  $140.00        $14.00             NaN                 1.3    1     0   
790  $148.70        $14.90             NaN                 9.3    0     0   

    Add to Cart  
624           0  
790           1  

[2 rows x 27 columns]

有2条交易数据缺失 Order Priority变量值,

从输出结果来看,这些缺失Description的交易数据,Region 都为North。为了验证猜想,我们增加筛选条件,看是否存在
 Order Priority 变量缺失且Region 不为North的数据。

In [10]:
original_data[(original_data['Order Priority'].isnull()) & ( original_data["Region"] != "North")]

Empty DataFrame
Columns: [Customer ID, Customer Name, Segment, City, State, Country, Region, Gender, Age, Education, Marital Status, Order ID, Order Date, Months, Ship Mode, Product Category, Product, Sales, Quantity, Discount, Profit, Shipping Cost, Order Priority, Browsing Time (min), Like, Share, Add to Cart]
Index: []

[0 rows x 27 columns]

### 评估重复数据

根据数据变量的含义来看,虽然 InvoiceNo、StockCode和CustomerID都是唯一标识符,但一次交易可能包含多件商
品,因此InvoiceNo 可以存在重复,不同交易可以包含同件商品,因此stockCode 可以存在重复,顾客可以进行多次交易或
下单多个商品,因此CustomerID 也可以存在重复。

那么针对此数据集,我们无需评估重复数据。


In [11]:
original_data["Customer ID"].duplicated()

0        False
1        False
2        False
3        False
4        False
         ...  
51284     True
51285     True
51286     True
51287     True
51288     True
Name: Customer ID, Length: 51289, dtype: bool

In [12]:
original_data[original_data.duplicated()]#说明咩有一行是完全一样的

Empty DataFrame
Columns: [Customer ID, Customer Name, Segment, City, State, Country, Region, Gender, Age, Education, Marital Status, Order ID, Order Date, Months, Ship Mode, Product Category, Product, Sales, Quantity, Discount, Profit, Shipping Cost, Order Priority, Browsing Time (min), Like, Share, Add to Cart]
Index: []

[0 rows x 27 columns]

In [13]:
original_data[original_data[["Customer ID","Customer Name","Education"]].duplicated()]

Customer ID    Customer Name      Segment           City          State  \
19        ED-0011       Johns Reed    Corporate         Taipei    Taipei City   
36        KE-0022      Holt Glocke    Corporate        Seattle     Washington   
39        RE-0018   Wong Macintyre     Consumer  New York City       New York   
63        ED-0011       Johns Reed    Corporate         Taipei    Taipei City   
67        EN-0059     Kelly Braden    Corporate        Detroit       Michigan   
...           ...              ...          ...            ...            ...   
51284    IN-00783       Welch Fein    Corporate        Baghdad        Baghdad   
51285    TT-00135  Martinez Arnett    Corporate         Kanpur  Uttar Pradesh   
51286    ON-00828     Mccoy Duston  Home Office       Portland         Oregon   
51287     RN-0016   Bentley Zypern     Consumer         Berlin         Berlin   
51288   RZ-001718  Mcclure Schwarz  Home Office      La Romana      La Romana   

                  Country        Region  Gender  Age         Education  ...  \
19                 Taiwan    North Asia    Male   43            Master  ...   
36          United States          West    Male   31          Bachelor  ...   
39          United States          East    Male   33          Bachelor  ...   
63                 Taiwan    North Asia    Male   43            Master  ...   
67          United States       Central  Female   46         Doctorate  ...   
...                   ...           ...     ...  ...               ...  ...   
51284                Iraq          EMEA  Female   20  Associate Degree  ...   
51285               India  Central Asia    Male   20       High School  ...   
51286       United States          West    Male   33            Master  ...   
51287             Germany       Central  Female   48         Doctorate  ...   
51288  Dominican Republic     Caribbean  Female   36            Master  ...   

          Sales Quantity Discount    Profit Shipping Cost Order Priority  \
19     $211.00         2     0.02  $122.60        $12.30        Critical   
36     $140.00         5     0.04   $32.00         $3.20          Medium   
39     $118.00         5     0.04   $14.40         $1.40        Critical   
63     $140.00         1     0.05   $53.00         $5.30        Critical   
67     $250.00         1     0.01  $167.50        $16.80            High   
...         ...      ...      ...       ...           ...            ...   
51284   $85.00         5     0.04   $17.00         $1.70          Medium   
51285   $85.00         1     0.03    $2.50         $0.20          Medium   
51286   $85.00         1     0.05    $0.80         $0.10          Medium   
51287   $85.00         3     0.04   $28.30         $2.80          Medium   
51288   $85.00         3     0.03   $28.30         $2.80          Medium   

      Browsing Time (min) Like Share Add to Cart  
19                   14.4    0     0           1  
36                   15.6    1     1           1  
39                   18.4    1     1           1  
63                    6.0    1     1           1  
67                   27.0    1     0           1  
...                   ...  ...   ...         ...  
51284                21.4    1     1           1  
51285                15.5    0     0           1  
51286                14.3    0     0           1  
51287                 8.4    1     1           1  
51288                25.8    1     1           1  

[50494 rows x 27 columns]

WWWWWW这个函数展示的结果出现里省略，怎么不省略
###如果你的 value_counts()输出结果被省略（显示为 ...），通常是因为 ​数据量较大，pandas 默认只显示部分结果。要 ​完整显示所有统计结果，可以通过以下方法解决：
方法一：###将结果保存到文本文件
#####original_data['Country'].value_counts().to_csv('country_counts.txt', header=True)
方法二：调整 pandas 显示设置（推荐）​​
在代码前设置 pd.set_option()，强制 pandas 显示所有行：
pd.set_option('display.max_rows', None)# None 表示无限制
original_data['Country'].value_counts()

### 评估不一致数据

不一致数据可能存在于Country 变量中,我们要查看是否存在多个不同值指代同一国家的情况。

In [14]:


#pd.set_option('display.max_rows', None)# None 表示无限制
original_data['Country'].value_counts()

Country
United States    9900
France           3995
Australia        3779
Mexico           3669
China            3328
                 ... 
Somalia            54
Martinique         53
Togo               50
Croatia            43
Syria              36
Name: count, Length: 84, dtype: int64

从Country 变量值来看,“USA”、"United States”均在表示美国,"United Kingdom"
示英国,因此应该在清洗步骤对这些值进行统一,只保留一个指代值。

"UK"

"U.K."均在表

### 评估无效或错误数据

可以通过DataFrame的 describe方法,对数值统计信息进行快速了解。

In [15]:
original_data.describe()

Age  Browsing Time (min)          Like         Share  \
count  51289.000000         51289.000000  51289.000000  51289.000000   
mean      31.315428            17.275973      0.567315      0.505274   
std       11.900634             9.674191      0.495453      0.499977   
min       18.000000             1.000000      0.000000      0.000000   
25%       22.000000             9.300000      0.000000      0.000000   
50%       28.000000            16.500000      1.000000      1.000000   
75%       37.000000            24.600000      1.000000      1.000000   
max       70.000000            40.000000      1.000000      1.000000   

        Add to Cart  
count  51289.000000  
mean       0.666732  
std        0.471386  
min        0.000000  
25%        0.000000  
50%        1.000000  
75%        1.000000  
max        1.000000

假设：从输出结果来看,Age和UnitPrice 存在负数,会对后续数值分析产生影响。

因此,我们先筛选出Age数值为负数的观察值,进一步评估其含义。

In [16]:
original_data[original_data["Age"]<0]

Empty DataFrame
Columns: [Customer ID, Customer Name, Segment, City, State, Country, Region, Gender, Age, Education, Marital Status, Order ID, Order Date, Months, Ship Mode, Product Category, Product, Sales, Quantity, Discount, Profit, Shipping Cost, Order Priority, Browsing Time (min), Like, Share, Add to Cart]
Index: []

[0 rows x 27 columns]

选出Age>30且，性别为Female，国家c开头的人

In [17]:
original_data[(original_data["Age"] < 30) & (original_data["Gender"]=="Female") & (original_data["Country"].str[0]=="C") ]

Customer ID       Customer Name      Segment              City  \
32        KI-0033     Garner Hirasaki     Consumer            Suzhou   
61        ON-0062   Sellers Stevenson     Consumer         Changchun   
75        CH-0076  Parsons Leinenbach    Corporate           Qingdao   
155      AK-00157           Wyatt Pak  Home Office            Ningbo   
166      AN-00168     Haley Wasserman     Consumer  Santiago de Cuba   
...           ...                 ...          ...               ...   
51098  PP-0041952          Tyler Kipp    Corporate            Zagreb   
51104    ES-00636         Dunn Cortes  Home Office            Bogotá   
51108    ER-00822          Combs Ober     Consumer             Arica   
51171   ON-001260      Nguyen Emerson     Consumer           Jingmen   
51233     KI-0033     Garner Hirasaki     Consumer            Suzhou   

                    State   Country      Region  Gender  Age  \
32                  Anhui     China  North Asia  Female   20   
61                  Jilin     China  North Asia  Female   23   
75               Shandong     China  North Asia  Female   25   
155              Zhejiang     China  North Asia  Female   26   
166      Santiago de Cuba      Cuba   Caribbean  Female   21   
...                   ...       ...         ...     ...  ...   
51098         Grad Zagreb   Croatia        EMEA  Female   20   
51104              Bogota  Colombia       South  Female   24   
51108  Arica y Parinacota     Chile       South  Female   19   
51171               Hubei     China  North Asia  Female   23   
51233               Anhui     China  North Asia  Female   20   

              Education  ...     Sales Quantity Discount    Profit  \
32          High School  ...   $72.00         4     0.05   $18.00    
61             Bachelor  ...  $114.00         3     0.01   $30.60    
75          High School  ...  $118.00         1     0.01   $36.80    
155            Bachelor  ...  $118.00         1     0.05   $32.10    
166         High School  ...   $72.00         1     0.03   $72.00    
...                 ...  ...       ...      ...      ...       ...   
51098  Associate Degree  ...  $224.00         5     0.03  $110.40    
51104          Bachelor  ...  $196.00         1     0.04  $108.20    
51108  Associate Degree  ...  $122.00         4     0.05   $17.60    
51171       High School  ...  $218.00         1     0.02  $133.60    
51233       High School  ...  $228.00         2     0.04  $129.80    

      Shipping Cost Order Priority Browsing Time (min) Like Share Add to Cart  
32           $1.80            High                21.5    0     0           1  
61           $3.10            High                 9.1    0     0           1  
75           $3.70            High                12.3    1     0           0  
155          $3.20        Critical                 2.1    0     0           0  
166          $7.20            High                37.7    0     0           0  
...             ...            ...                 ...  ...   ...         ...  
51098       $11.00          Medium                20.6    1     0           1  
51104       $10.80          Medium                10.6    1     1           1  
51108        $1.80          Medium                10.9    1     1           1  
51171       $13.40          Medium                34.9    1     1           1  
51233       $13.00          Medium                 9.7    1     1           1  

[1630 rows x 27 columns]

# 清理数据
假设：代做清单

根据前面评估部分得到的结论,我们需要进行的数据清理包括:

· 把InvoiceDate变量的数据类型转换为为日期时间

· 把CustomerID变量的数据类型转换为字符串

· 把Description变量缺失的观察值删除

· 把Country 变量值“USA”替换为"United States"

· 把Country 变量值“UK”、“U.K.”替换为"United Kingdom"

· 把Quantity变量值为负数的观察值删除

■ 把UnitPrice 变量值为负数的观察值删除

为了区分开经过清理的数据和原始的数据,我们创建新的变量cleaned_data,让它为original_data复制出的副本。我们
之后的清理步骤都将被运用在cleaned_data上。

In [18]:
cleaned_data = original_data.copy()
cleaned_data.head()

Customer ID    Customer Name      Segment        City       State  \
0      LS-001     Lane Daniels     Consumer    Brisbane  Queensland   
1      IZ-002    Alvarado Kriz  Home Office      Berlin      Berlin   
2      EN-003       Moon Weien     Consumer     Porirua  Wellington   
3      AN-004  Sanchez Bergman    Corporate       Kabul       Kabul   
4      ON-005     Rowe Jackson    Corporate  Townsville  Queensland   

       Country        Region Gender  Age         Education  ...     Sales  \
0    Australia       Oceania   Male   22  Associate Degree  ...  $140.00    
1      Germany       Central   Male   32          Bachelor  ...  $211.00    
2  New Zealand       Oceania   Male   21       High School  ...  $117.00    
3  Afghanistan  Central Asia   Male   22          Bachelor  ...  $118.00    
4    Australia       Oceania   Male   28  Associate Degree  ...  $250.00    

  Quantity Discount    Profit Shipping Cost Order Priority  \
0        2     0.05   $46.00         $4.60          Medium   
1        3     0.03  $112.00        $11.20          Medium   
2        5     0.01   $31.20         $3.10        Critical   
3        2     0.05   $26.20         $2.60            High   
4        1     0.04  $160.00        $16.00        Critical   

  Browsing Time (min) Like Share Add to Cart  
0                14.7    1     1           1  
1                15.0    0     0           1  
2                19.9    1     1           1  
3                15.8    1     1           1  
4                18.1    1     1           1  

[5 rows x 27 columns]

把 InvoiceDate 变量的数据类型转换为为日期时间类型:

In [19]:
cleaned_data["Browsing Time (min)"] = pd.to_datetime(cleaned_data["Browsing Time (min)"])
cleaned_data["Browsing Time (min)"]

0       1970-01-01 00:00:00.000000014
1       1970-01-01 00:00:00.000000015
2       1970-01-01 00:00:00.000000019
3       1970-01-01 00:00:00.000000015
4       1970-01-01 00:00:00.000000018
                     ...             
51284   1970-01-01 00:00:00.000000021
51285   1970-01-01 00:00:00.000000015
51286   1970-01-01 00:00:00.000000014
51287   1970-01-01 00:00:00.000000008
51288   1970-01-01 00:00:00.000000025
Name: Browsing Time (min), Length: 51289, dtype: datetime64[ns]

把Discount 变量的数据类型转换为字符串:

In [20]:
cleaned_data["Discount"] = cleaned_data["Discount"].astype(str)
cleaned_data["Discount"]

0        0.05
1        0.03
2        0.01
3        0.05
4        0.04
         ... 
51284    0.04
51285    0.03
51286    0.05
51287    0.04
51288    0.03
Name: Discount, Length: 51289, dtype: object

把Discount 变量值结尾的“.0”删除:

In [21]:
cleaned_data["Discount"]=cleaned_data["Discount"].str.slice(0,-2)
cleaned_data["Discount"]

0        0.
1        0.
2        0.
3        0.
4        0.
         ..
51284    0.
51285    0.
51286    0.
51287    0.
51288    0.
Name: Discount, Length: 51289, dtype: object

把Order Priority 变量缺失的观察值删除,并查看删除后该列空缺值个数和:

In [22]:
# ​如果是 Series​（一维数据）：
# •cleaned_data = cleaned_data.dropna(subset=["Order Priority"])
# cleaned_data["Order Priority"].isnull().sum()
# Series的 dropna()​不支持 subset，因为它只能对整个 Series 删除缺失值。
cleaned_data = pd.DataFrame(cleaned_data)  # 转换为DataFrame
cleaned_data = cleaned_data.dropna(subset=["Order Priority"])
cleaned_data["Order Priority"].isnull().sum()

0

In [23]:
print(type(cleaned_data))  # 输出类型，例如 <class 'pandas.core.frame.DataFrame'> 或 <class 'pandas.core.series.Series'>

<class 'pandas.core.frame.DataFrame'>


In [24]:
print(type(cleaned_data)) 

<class 'pandas.core.frame.DataFrame'>


In [25]:
#pip install --upgrade pandas

In [26]:
cleaned_data = pd.DataFrame(cleaned_data)
print(type(cleaned_data))

<class 'pandas.core.frame.DataFrame'>


把Country 变量值“US”替换为“United States”,并检查替换后“US”变量值个数:

In [27]:
cleaned_data = pd.DataFrame(cleaned_data)
cleaned_data= cleaned_data["Country"].replace({"United States":"US"})
# len(cleaned_data[cleaned_data["Country"]=="US"])

In [28]:
cleaned_data = pd.DataFrame(cleaned_data)
len(cleaned_data["Country"]=="US")#返回一个布尔值Series的长度

51287

In [29]:
cleaned_data = pd.DataFrame(cleaned_data)
(cleaned_data["Country"]=="US").sum()#返回实际满足条件的记录数量

9900

In [30]:
cleaned_data = pd.DataFrame(cleaned_data)
len(cleaned_data[cleaned_data["Country"]=="US"])#返回实际满足条件的记录数量

9900

In [31]:
cleaned_data = pd.DataFrame(cleaned_data)
(cleaned_data["Country"]=="United States").sum()

0

In [32]:
cleaned_data["Country"]

0                 Australia
1                   Germany
2               New Zealand
3               Afghanistan
4                 Australia
                ...        
51284                  Iraq
51285                 India
51286                    US
51287               Germany
51288    Dominican Republic
Name: Country, Length: 51287, dtype: object

In [33]:
cleaned_data = pd.DataFrame(cleaned_data)
print(cleaned_data.columns)

Index(['Country'], dtype='object')


In [34]:
print(type(cleaned_data))

<class 'pandas.core.frame.DataFrame'>


把Country 变量值“Australia”“India”,替换为“ New Zealand”,并检查替换后“Australia”“India”变量值个数:

In [35]:
cleaned_data = cleaned_data.replace(["Australia","India"],"New Zealand")
print((cleaned_data["Country"] == "Australia").sum())
print((cleaned_data["Country"] == "India").sum())


0
0


In [36]:
(cleaned_data["Country"] == "New Zealand").sum()

6137

In [37]:
cleaned_data = original_data.copy()
cleaned_data.head()

Customer ID    Customer Name      Segment        City       State  \
0      LS-001     Lane Daniels     Consumer    Brisbane  Queensland   
1      IZ-002    Alvarado Kriz  Home Office      Berlin      Berlin   
2      EN-003       Moon Weien     Consumer     Porirua  Wellington   
3      AN-004  Sanchez Bergman    Corporate       Kabul       Kabul   
4      ON-005     Rowe Jackson    Corporate  Townsville  Queensland   

       Country        Region Gender  Age         Education  ...     Sales  \
0    Australia       Oceania   Male   22  Associate Degree  ...  $140.00    
1      Germany       Central   Male   32          Bachelor  ...  $211.00    
2  New Zealand       Oceania   Male   21       High School  ...  $117.00    
3  Afghanistan  Central Asia   Male   22          Bachelor  ...  $118.00    
4    Australia       Oceania   Male   28  Associate Degree  ...  $250.00    

  Quantity Discount    Profit Shipping Cost Order Priority  \
0        2     0.05   $46.00         $4.60          Medium   
1        3     0.03  $112.00        $11.20          Medium   
2        5     0.01   $31.20         $3.10        Critical   
3        2     0.05   $26.20         $2.60            High   
4        1     0.04  $160.00        $16.00        Critical   

  Browsing Time (min) Like Share Add to Cart  
0                14.7    1     1           1  
1                15.0    0     0           1  
2                19.9    1     1           1  
3                15.8    1     1           1  
4                18.1    1     1           1  

[5 rows x 27 columns]

In [38]:
cleaned_data.info#cleaned_data.info不等于cleaned_data.info（）

<bound method DataFrame.info of       Customer ID    Customer Name      Segment        City          State  \
0          LS-001     Lane Daniels     Consumer    Brisbane     Queensland   
1          IZ-002    Alvarado Kriz  Home Office      Berlin         Berlin   
2          EN-003       Moon Weien     Consumer     Porirua     Wellington   
3          AN-004  Sanchez Bergman    Corporate       Kabul          Kabul   
4          ON-005     Rowe Jackson    Corporate  Townsville     Queensland   
...           ...              ...          ...         ...            ...   
51284    IN-00783       Welch Fein    Corporate     Baghdad        Baghdad   
51285    TT-00135  Martinez Arnett    Corporate      Kanpur  Uttar Pradesh   
51286    ON-00828     Mccoy Duston  Home Office    Portland         Oregon   
51287     RN-0016   Bentley Zypern     Consumer      Berlin         Berlin   
51288   RZ-001718  Mcclure Schwarz  Home Office   La Romana      La Romana   

                  Country      

In [39]:
cleaned_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51289 entries, 0 to 51288
Data columns (total 27 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Customer ID          51289 non-null  object 
 1   Customer Name        51289 non-null  object 
 2   Segment              51289 non-null  object 
 3   City                 51289 non-null  object 
 4   State                51289 non-null  object 
 5   Country              51289 non-null  object 
 6   Region               51215 non-null  object 
 7   Gender               51289 non-null  object 
 8   Age                  51289 non-null  int64  
 9   Education            51289 non-null  object 
 10  Marital Status       51289 non-null  object 
 11  Order ID             51289 non-null  object 
 12  Order Date           51289 non-null  object 
 13  Months               51289 non-null  object 
 14  Ship Mode            51289 non-null  object 
 15  Product Category     51289 non-null 

把Quantity 变量值为负数的观察值删除,并检查替换后Quantity 变量值为负数的个数:

In [40]:
cleaned_data = cleaned_data[cleaned_data["Like"]>=0]
(cleaned_data["Like"] < 0).sum()

0

In [41]:
cleaned_data["Quantity"] = cleaned_data["Quantity"].fillna(0)

# 保存清理后的数据

完成数据清理后,把干净整卉的数据保存到新的文件里,文件名为e_commerce_cleaned.csv

In [42]:
cleaned_data.to_csv("e_commerce_cleaned.csv")

In [43]:
cleaned_data = pd.DataFrame(cleaned_data)
pd.read_csv("e_commerce_cleaned.csv")

<ipython-input-43-36afbf251daa>:2: DtypeWarning: Columns (19,20) have mixed types. Specify dtype option on import or set low_memory=False.
  pd.read_csv("e_commerce_cleaned.csv")


Unnamed: 0 Customer ID    Customer Name      Segment        City  \
0               0      LS-001     Lane Daniels     Consumer    Brisbane   
1               1      IZ-002    Alvarado Kriz  Home Office      Berlin   
2               2      EN-003       Moon Weien     Consumer     Porirua   
3               3      AN-004  Sanchez Bergman    Corporate       Kabul   
4               4      ON-005     Rowe Jackson    Corporate  Townsville   
...           ...         ...              ...          ...         ...   
51284       51284    IN-00783       Welch Fein    Corporate     Baghdad   
51285       51285    TT-00135  Martinez Arnett    Corporate      Kanpur   
51286       51286    ON-00828     Mccoy Duston  Home Office    Portland   
51287       51287     RN-0016   Bentley Zypern     Consumer      Berlin   
51288       51288   RZ-001718  Mcclure Schwarz  Home Office   La Romana   

               State             Country        Region  Gender  Age  ...  \
0         Queensland           Australia       Oceania    Male   22  ...   
1             Berlin             Germany       Central    Male   32  ...   
2         Wellington         New Zealand       Oceania    Male   21  ...   
3              Kabul         Afghanistan  Central Asia    Male   22  ...   
4         Queensland           Australia       Oceania    Male   28  ...   
...              ...                 ...           ...     ...  ...  ...   
51284        Baghdad                Iraq          EMEA  Female   20  ...   
51285  Uttar Pradesh               India  Central Asia    Male   20  ...   
51286         Oregon       United States          West    Male   33  ...   
51287         Berlin             Germany       Central  Female   48  ...   
51288      La Romana  Dominican Republic     Caribbean  Female   36  ...   

          Sales Quantity Discount    Profit Shipping Cost Order Priority  \
0      $140.00         2     0.05   $46.00         $4.60          Medium   
1      $211.00         3     0.03  $112.00        $11.20          Medium   
2      $117.00         5     0.01   $31.20         $3.10        Critical   
3      $118.00         2     0.05   $26.20         $2.60            High   
4      $250.00         1     0.04  $160.00        $16.00        Critical   
...         ...      ...      ...       ...           ...            ...   
51284   $85.00         5     0.04   $17.00         $1.70          Medium   
51285   $85.00         1     0.03    $2.50         $0.20          Medium   
51286   $85.00         1     0.05    $0.80         $0.10          Medium   
51287   $85.00         3     0.04   $28.30         $2.80          Medium   
51288   $85.00         3     0.03   $28.30         $2.80          Medium   

      Browsing Time (min) Like Share Add to Cart  
0                    14.7    1     1           1  
1                    15.0    0     0           1  
2                    19.9    1     1           1  
3                    15.8    1     1           1  
4                    18.1    1     1           1  
...                   ...  ...   ...         ...  
51284                21.4    1     1           1  
51285                15.5    0     0           1  
51286                14.3    0     0           1  
51287                 8.4    1     1           1  
51288                25.8    1     1           1  

[51289 rows x 28 columns]

In [45]:
cleaned_data.to_csv("e_commerce_cleaned.csv",index = False)#去索引储存

In [46]:
cleaned_data = pd.DataFrame(cleaned_data)
pd.read_csv("e_commerce_cleaned.csv")

<ipython-input-46-36afbf251daa>:2: DtypeWarning: Columns (18,19) have mixed types. Specify dtype option on import or set low_memory=False.
  pd.read_csv("e_commerce_cleaned.csv")


Customer ID    Customer Name      Segment        City          State  \
0          LS-001     Lane Daniels     Consumer    Brisbane     Queensland   
1          IZ-002    Alvarado Kriz  Home Office      Berlin         Berlin   
2          EN-003       Moon Weien     Consumer     Porirua     Wellington   
3          AN-004  Sanchez Bergman    Corporate       Kabul          Kabul   
4          ON-005     Rowe Jackson    Corporate  Townsville     Queensland   
...           ...              ...          ...         ...            ...   
51284    IN-00783       Welch Fein    Corporate     Baghdad        Baghdad   
51285    TT-00135  Martinez Arnett    Corporate      Kanpur  Uttar Pradesh   
51286    ON-00828     Mccoy Duston  Home Office    Portland         Oregon   
51287     RN-0016   Bentley Zypern     Consumer      Berlin         Berlin   
51288   RZ-001718  Mcclure Schwarz  Home Office   La Romana      La Romana   

                  Country        Region  Gender  Age         Education  ...  \
0               Australia       Oceania    Male   22  Associate Degree  ...   
1                 Germany       Central    Male   32          Bachelor  ...   
2             New Zealand       Oceania    Male   21       High School  ...   
3             Afghanistan  Central Asia    Male   22          Bachelor  ...   
4               Australia       Oceania    Male   28  Associate Degree  ...   
...                   ...           ...     ...  ...               ...  ...   
51284                Iraq          EMEA  Female   20  Associate Degree  ...   
51285               India  Central Asia    Male   20       High School  ...   
51286       United States          West    Male   33            Master  ...   
51287             Germany       Central  Female   48         Doctorate  ...   
51288  Dominican Republic     Caribbean  Female   36            Master  ...   

          Sales Quantity Discount    Profit Shipping Cost Order Priority  \
0      $140.00         2     0.05   $46.00         $4.60          Medium   
1      $211.00         3     0.03  $112.00        $11.20          Medium   
2      $117.00         5     0.01   $31.20         $3.10        Critical   
3      $118.00         2     0.05   $26.20         $2.60            High   
4      $250.00         1     0.04  $160.00        $16.00        Critical   
...         ...      ...      ...       ...           ...            ...   
51284   $85.00         5     0.04   $17.00         $1.70          Medium   
51285   $85.00         1     0.03    $2.50         $0.20          Medium   
51286   $85.00         1     0.05    $0.80         $0.10          Medium   
51287   $85.00         3     0.04   $28.30         $2.80          Medium   
51288   $85.00         3     0.03   $28.30         $2.80          Medium   

      Browsing Time (min) Like Share Add to Cart  
0                    14.7    1     1           1  
1                    15.0    0     0           1  
2                    19.9    1     1           1  
3                    15.8    1     1           1  
4                    18.1    1     1           1  
...                   ...  ...   ...         ...  
51284                21.4    1     1           1  
51285                15.5    0     0           1  
51286                14.3    0     0           1  
51287                 8.4    1     1           1  
51288                25.8    1     1           1  

[51289 rows x 27 columns]